File: app/src/main/java/com/yuyh/jsonviewer/MainActivity.java
package com.yuyh.jsonviewer;

import android.os.Bundle;

import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.HorizontalScrollView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.yuyh.jsonviewer.library.JsonRecyclerView;
import com.yuyh.jsonviewer.library.adapter.JsonViewerAdapter;

import java.io.IOException;
import java.io.InputStream;

public class MainActivity extends AppCompatActivity {

    private JsonRecyclerView mRecyclewView;
    private HorizontalScrollView mHScrollView;

    private boolean isExpanded = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mHScrollView = findViewById(R.id.hsv);

        mRecyclewView = findViewById(R.id.rv_json);
        mRecyclewView.setScaleEnable(true);
        mRecyclewView.addOnItemTouchListener(new RecyclerView.OnItemTouchListener() { // 避免双指缩放与上下左右滑动冲突
            @Override
            public boolean onInterceptTouchEvent(RecyclerView rv, MotionEvent event) {
                switch (event.getAction() & event.getActionMasked()) {
                    case MotionEvent.ACTION_DOWN:
                        break;
                    case MotionEvent.ACTION_UP:
                        break;
                    case MotionEvent.ACTION_POINTER_UP:
                        mHScrollView.requestDisallowInterceptTouchEvent(false);
                        break;
                    case MotionEvent.ACTION_POINTER_DOWN:
                        mHScrollView.requestDisallowInterceptTouchEvent(true);
                        break;

                    case MotionEvent.ACTION_MOVE:
                        break;
                }
                return false;
            }

            @Override
            public void onTouchEvent(RecyclerView rv, MotionEvent e) {

            }

            @Override
            public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {

            }
        });

        new Thread(new Runnable() {
            @Override
            public void run() {
                InputStream is = null;
                try {
                    is = getAssets().open("demo_smaller.json");
                    int lenght = is.available();
                    byte[] buffer = new byte[lenght];
                    is.read(buffer);
                    final String result = new String(buffer, "utf8");
                    is.close();

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            mRecyclewView.bindJson(result);
                        }
                    });
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }).start();

        View toggle = findViewById(R.id.toggle);
        toggle.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (isExpanded) {
                    mRecyclewView.collapseAll();
                } else {
                    mRecyclewView.expandAllToDepth(2);
                }
                isExpanded = !isExpanded;
            }
        });
    }
}

File: app/src/main/java/com/yuyh/jsonviewer/MainActivity.kt
package com.yuyh.jsonviewer

import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.yuyh.jsonviewer.library.adapter.JsonViewerAdapter
import kotlinx.android.synthetic.main.activity_main.*
import java.io.IOException
import java.io.InputStream

class MainActivity : AppCompatActivity() {
    private var isExpanded = false
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        rv_json.depth = 3
        rv_json.setScaleEnable(true)
        rv_json.addOnItemTouchListener(object : RecyclerView.OnItemTouchListener {

            override fun onInterceptTouchEvent(rv: RecyclerView, event: MotionEvent): Boolean {
                when (event.action and event.actionMasked) {
                    MotionEvent.ACTION_DOWN -> { }
                    MotionEvent.ACTION_UP -> { }
                    MotionEvent.ACTION_POINTER_UP -> hsv.requestDisallowInterceptTouchEvent(false)
                    MotionEvent.ACTION_POINTER_DOWN -> hsv.requestDisallowInterceptTouchEvent(true)
                    MotionEvent.ACTION_MOVE -> { }
                }
                return false
            }

            override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {}
            override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}
        })
        Thread(Runnable {
            var `is`: InputStream? = null
            try {
                `is` = assets.open("demo_smaller.json")
                val lenght = `is`.available()
                val buffer = ByteArray(lenght)
                `is`.read(buffer)
                val result = String(buffer, Charsets.UTF_8)
                `is`.close()
                runOnUiThread { rv_json.bindJson(result) }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }).start()

        val toggle = findViewById<View>(R.id.toggle)
        toggle.setOnClickListener {
            if (isExpanded) {
                rv_json.collapseAll()
            } else {
                rv_json.expandAllToDepth(2)
            }
            isExpanded = !isExpanded
        }
    }
}